### PR TITLE
Don't allocate when creating an empty BTree

### LINKED ISSUE #50266: Make BTreeMap::new() not allocate
Currently, `BTreeMap::new()` allocates an empty root node. This is in spite of all of our other collections managing to avoid this. In #50240 it was demonstrated that there's massive wins to be had by avoiding allocating here. In the past we avoided implementing this because it simplified the code of BTreeMap a lot (which is already pretty gnarly code).

I see two ways to do this

* Wrap everything in Options, adding extra branches on accesses to iterators and the map itself.
* Do lots of clever shit, introducing minimal branches on access.

Some examples of clever shit:

* The core Range iterator primitive can always always be initialized with two `NonZero::dangling()` pointers and `len = 0`. The existing branches will then do the correct thing. (this is how empty slices generally work)

* By changing the layout of LeafNode to have the metadata *first*, we can statically allocate a `LeafNode<u64, u64>` (say) for all empty instances of BTreeMap to share. get/remove queries will then naturally work correctly without extra branches. See [thin-vec](https://github.com/Gankro/thin-vec/blob/master/src/lib.rs#L251-L261) for an example of doing this. An extra branch will probably need to be added to `insert` to check if we're pointing at the static singleton; not a big deal.

### PR DESCRIPTION
**Pull request by porglezomp**
Following the discussion in #50266, this adds a static instance of `LeafNode` that empty BTrees point to, and then replaces it on `insert`, `append`, and `entry`. This avoids allocating for empty maps.

Fixes #50266

r? @Gankro

### CODE REVIEW

**Comment by porglezomp** [comment]
I was having problems getting tests to run locally and got impatient, so I guess I'm going to be fixing that. In order to make this more robust for future maintainers, I'm thinking of adding some type-state to Root so that you can't do write operations without first verifying that it's not the shared static root.

For example, `Root<K, V, marker::SharedOrOwned>` and `Root<K, V, marker::Owned>`, storing the former, and then having the `insert` functions begin by converting it into the latter.

File: src/liballoc/btree/node.rs
Comment (Gankro) [inline src/liballoc/btree/node.rs]: why not hoist the parent pointer too?

File: src/liballoc/btree/node.rs
Comment (Gankro) [inline src/liballoc/btree/node.rs]: is there any reason to use ptr::eq and not just `==` ?

Comment (Gankro) [reply src/liballoc/btree/node.rs]: (Also is the as_ref() necessary..?)

File: src/liballoc/btree/node.rs
Comment (Gankro) [inline src/liballoc/btree/node.rs]: I would rather use the `new_unchecked` constructor for NonNull.

**Comment by Gankro** [comment]
Wow this turned out really simple!

That said I think you need to be careful with the empty singleton. If any BTreeMap with non-zst K/V tries to take a pointer to the first value or last key, this will construct an out-of-allocation pointer using ptr::offset (things like array indexing all lower to that), which is UB. In addition types with a higher alignment that 8 (or 4 on 32-bit) will access out-of-allocation if they try to take a pointer to the first key.

You should take a quick audit for that kinda stuff. For instance any call to keys() will create such a pointer. This is why thin-vec has this awful mess: https://github.com/Gankro/thin-vec/blob/master/src/lib.rs#L99-L113

File: src/liballoc/btree/node.rs
Comment (porglezomp) [reply src/liballoc/btree/node.rs]: `ptr::eq` just to be explicit about not deref-ing, but that's not necessary if they're raw pointers, and the `as_ref` was because I expected to compare the the reference not the raw pointer. I'll clean that up.

File: src/liballoc/btree/node.rs
Comment (porglezomp) [reply src/liballoc/btree/node.rs]: I was trying to figure out what the ideal choice for alignment/packing would be, but this is probably more general (so it can see that it's null?)

File: src/liballoc/btree/node.rs
Comment (porglezomp) [reply src/liballoc/btree/node.rs]: Oh wait, yes, `as_ref()` is necessary because `as_ptr()` returns a `NonNull<_>`, so `ptr::eq` lets you avoid an extra pair of coercions.

I think `.as_ptr().as_ptr()` is better than `.as_ptr().as_ref()` though since then we don't need `unsafe`.

**Comment by Gankro** [comment]
I think you can probably just get away with making key()/vals() do the `if padding { ... }` check, I do in thin-vec, but definitely check my work on that.

**Comment by porglezomp** [comment]
So, presumably the logic behind that `if padding { … }` is that it will get constant-folded, and then you'll just use the potentially-out-of-bounds version for unaligned data? I think the issue here is that, say, you have `K = [u8; 16]`, then `values()` will be out-of-bounds even though the keys and vals are perfectly aligned.

I can check if it's pointing to the shared root and then use `NonNull::dangling()` to make the slices? That's a branch that won't get constant folded, but can probably become a cmov at least?

File: src/liballoc/btree/node.rs
Comment (porglezomp) [reply src/liballoc/btree/node.rs]: See the `Drop` impl on `IntoIter`, it calls `leaf_node.deallocate_and_ascend()` on the root, so it will always try to deallocate the root node. I should move that check into that drop impl though.

Comment (arthurprs) [reply src/liballoc/btree/node.rs]: I see, thanks for the explanation.

File: src/liballoc/btree/node.rs
Comment (arthurprs) [inline src/liballoc/btree/node.rs]: Doesn't need to be unsafe

File: src/liballoc/btree/node.rs
Comment (arthurprs) [reply src/liballoc/btree/node.rs]: It might be cleaner to move this into the IntoIter Drop though, even if it takes another public method. It looks cleaner and `map.rs` is already aware of the shared root anyway.

**Comment by shepmaster** [comment]
Ping from triage, @Gankro — looks like there are new commits to review!

**Comment by shepmaster** [comment]
But the PR *is* marked as WIP, so moving back to waiting on author.

**Comment by porglezomp** [comment]
@Gankro CI issue is cleared up, and I squashed some of the commits that went back and forth (like swapping between `()` and `u64`).

**Comment by porglezomp** [comment]
I believe I could fix the `entry` allocating without inserting in two possible ways, neither one particularly appealing.

The first means making the `Entry` struct larger by adding an additional pointer to the `VacantEntry` struct, I'm not sure how appealing that is given that it's currently enum discriminant + 4 pointers + `K` large.

The second adds no size but is a horrendous scary type abuse (maybe UB? are you allowed to cast pointers to completely unrelated types?) making the `root` pointer in the `NodeRef` point to the `BTreeMap<K, V>` instead of its contained `Root<K, V>`, so that we can allocate when we need to.

**Comment by porglezomp** [comment]
Maybe that change should be saved for a separate PR though?

**Comment by Gankro** [comment]
Yes we can prevent Entry from allocating as a follow-up.

Also Rust has no notion of strict-aliasing/type-based-aliasing. As long as the layouts match you're golden.

File: src/liballoc/btree/map.rs
Comment (Gankro) [inline src/liballoc/btree/map.rs]: Could you abstract this pattern into a method? `ensure_root_is_owned()` or something?

File: src/liballoc/btree/node.rs
Comment (Gankro) [inline src/liballoc/btree/node.rs]: Is this even possible? It should literally be in rodata...

File: src/liballoc/btree/node.rs
Comment (Gankro) [inline src/liballoc/btree/node.rs]: ✨types

**Review by Gankra — CHANGES_REQUESTED** [review]
This looks great!

Just one minor nit, and I think this is good to go?

Although we should probably benchmark stuff...

File: src/liballoc/btree/node.rs
Comment (porglezomp) [reply src/liballoc/btree/node.rs]: Oh, right, that'd be why I was getting segfaults when it got modified. I'll remove this.

**Comment by Gankro** [comment]
Ok, r=me on the actual changes, but I'd like to get some profiles on this change before we actually land it, since this is ultimately an optimization.

**Review by Gankra — APPROVED** [review]

**Comment by porglezomp** [comment]
I've run the CPU benchmarks locally, I still seem to be having trouble getting the memory benchmarks working, but the CPU benchmarks see a uniform improvement on all fronts, from and average 0.4% improvement up to 2.9% improvement. The best individual benchmark results are 4.4% better, and only one sample is worse. `style-servo` clean incremental is 16% worse, which seems likely to be caused by something else on the machine? I'll try running them again.

**Comment by nnethercote** [comment]
`style-servo` clean incremental has enormous variation; ignore it!

Do your results look similar to those in https://github.com/rust-lang/rust/pull/50491#issuecomment-386951031 ? That's what I would expect.

**Comment by Gankro** [comment]
@bors-servo r+

Sounds like we're good to land this, great work!

**Comment by Gankro** [comment]
oh right

**Comment by porglezomp** [comment]
r=gankro

**Comment by nnethercote** [comment]
Nice work. I just looked at the code... I made the right decision in not attempting this myself :)

**Comment by nnethercote** [comment]
Oh, and once this lands, I'm happy to remove LazyBTreeMap.

**Comment by nnethercote** [comment]
@Gankro: this isn't in Homu's queue, apparently because an sufficiently privileged person hasn't approved it...

**Comment by Gankro** [comment]
Yeah i was assuming one would roll by.......

**Comment by kennytm** [comment]
@bors retry 

This is odd

**Pull request merged**
